Transfer of the smart contract notion docs (#7)
aragon/builders-portal: docs/core/02-how-to-guides/01-plugin-development/01-implemention/01-plugin-types.md

---
title: Plugin Types
---

## Choosing the Base Contract for Your Plugin

In this section, we will learn about the interfaces or base contracts to inherit from when developing a plugin.

> Depending on the use-case of your plugin, you might want it to be

Reviewer: ```suggestion
Depending on the use-case of your plugin, you may need it to be:
```

- upgradeable or non-upgradeable
- deployed by a specific deployment method
- compatible with meta transactions

:::caution

> Upgradeable plugin contracts (such as `PluginUpgradeable`, `PluginUUPSUpgradeable` implementations) must reserve storage slots by defining a storage gap. If this is not done properly and/or variables are rearranged in the upgrade, the storage can become corrupted rendering it inaccessible and resulting in the loss of funds.

Reviewer: Great add-on! I'd say let's add a link here to "storage gap" for someone who doesn't know what that is.

Author: Provided link to [OZs storage gap article](https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps).

:::

### Upgradeability & Deployment

Upgradebility and the deployment method of a plugin contract go hand in hand.

> The motivation to upgrade smart contracts is nicely summarized by OpenZepplin:

Reviewer: ```suggestion
The motivation behind upgrading smart contracts is nicely summarized by OpenZepplin:
```

> - change the rules enforced by it.
>
> Here’s what you’d need to do to fix a bug in a contract you cannot upgrade:
>
> 1. Deploy a new version of the contract
> 2. Manually migrate all state from the old one contract to the new one (which can be very expensive in terms of gas fees!)
> 3. Update all contracts that interacted with the old contract to use the address of the new one
> 4. Reach out to all your users and convince them to start using the new deployment (and handle both contracts being used simultaneously, as users are slow to migrate
>
> _source: [OpenZepplin: What's in an upgrade](https://docs.openzeppelin.com/learn/upgrading-smart-contracts#whats-in-an-upgrade)_

With upgradeable smart contracts, you can modify their code while keep using or even extending the storage (see the guide [Writing Upgradeable Contracts](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable) by OpenZepplin).

To enable upgradeable smart contracts (as well as cheap contract clones), the proxy pattern is used.

Depending on your upgradeability requirements and the deployment method you choose, you can also greatly reduce the gas costs to distribute your plugin.

> However, the upgradeability and deployment method can also introduce caveats during [the plugin setup](docs/core/01-how-it-works/02-the-dao-framework/02-plugin-marketplace/04-plugin-setup.md), especially for updating from an older version to a new one.

Reviewer: ```suggestion
However, the upgradeability and deployment method can introduce caveats during [the plugin setup](docs/core/01-how-it-works/02-the-dao-framework/02-plugin-marketplace/04-plugin-setup.md), especially when updating from an older version to a new one.
```

> The following table presents an overview of the different deployment methods and their benefits and drawbacks:

Reviewer: ```suggestion
The following table presents an overview of the different deployment methods with each of their benefits and drawbacks:
```

| plugin updating       | <span class="table-cell-negative">limited</span> | <span class="table-cell-negative">limited</span>  | <span class="table-cell-positive">easy</span>    | <span class="table-cell-positive">easy</span>    |

Accordingly, we recommend the UUPS proxy method for developing easily updatable Aragon Plugins and minimal clones for those, where the availability of the storage after the update is secondary / not needed.

To help you with developing and deploying the plugin within the Aragon infrastructure, we provide plugin interfaces and setup contracts for [the OpenZepplin Contracts Proxies](https://docs.openzeppelin.com/contracts/4.x/api/proxy) that you can inherit from besides the standard abstract `Plugin` contract

- `PluginClones`
- `PluginTransparentUpgradeable`
- `PluginUUPSUpgradeable`

#### Caveats of non-upgradeable Plugins

Aragon plugins using non-upgradeable smart contracts can be cheap to deploy (i.e., using clones) but are **limited when it comes to updating**.

> Updating, in distinction from upgrading, we call the Aragon DAO Framework specific process of switching from an older plugin version to a newer one.

Reviewer: ```suggestion
Updating, in distinction from upgrading, will call the aragonOS' internal process for switching from an older plugin version to a newer one.
```
